### ISSUE TITLE
Modulesd is killed when fetching NVD for Vulnerability Detector

### ISSUE THREAD

**Issue by mikykeane** [wazuh/wazuh#4398]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.11.0 | Wazuh Manager | Manager | Packages | CentOS 7 |

We have encountered an issue. In an AWS environment, with 1Gb of RAM and 1 CPU with these characteristics: 
```[root@wazuh-manager centos]# cat /proc/cpuinfo 
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 63
model name	: Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz
stepping	: 2
microcode	: 0x43
cpu MHz		: 2400.172
cache size	: 30720 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt
bogomips	: 4800.18
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

Whenever we enable NVD in `<vulnerability_detector>` the daemon `modulesd` is killed. We have isolated the problem only to NVD. It correctly fetches and downloads the database from Red Hat and runs without issue as long as NVD is disabled.

These are some of the logs we have found. In `/var/log/messages` 

```
Dec 26 14:42:15 wazuh-manager kernel: wazuh-modulesd invoked oom-killer: gfp_mask=0x201da, order=0, oom_score_adj=0
Dec 26 14:42:15 wazuh-manager kernel: wazuh-modulesd cpuset=/ mems_allowed=0
Dec 26 14:42:15 wazuh-manager kernel: CPU: 0 PID: 8521 Comm: wazuh-modulesd Not tainted 3.10.0-957.1.3.el7.x86_64 #1
Dec 26 14:42:15 wazuh-manager kernel: [ 8515]     0  8515   316060   167062     360        0             0 wazuh-modulesd
Dec 26 14:42:15 wazuh-manager kernel: Out of memory: Kill process 8515 (wazuh-modulesd) score 641 or sacrifice child
Dec 26 14:42:15 wazuh-manager kernel: Killed process 8515 (wazuh-modulesd) total-vm:1264240kB, anon-rss:668248kB, file-rss:0kB, shmem-rss:0kB
Dec 26 14:42:15 wazuh-manager kernel: [ 8519]     0  8515   316060   167097     360        0             0 wazuh-modulesd
Dec 26 14:42:15 wazuh-manager kernel: Out of memory: Kill process 8519 (wazuh-modulesd) score 641 or sacrifice child
Dec 26 14:42:15 wazuh-manager kernel: Killed process 8519 (wazuh-modulesd) total-vm:1264240kB, anon-rss:668388kB, file-rss:0kB, shmem-rss:0kB
```

When executing `/var/ossec/bin/wazuh-modulesd -ddf`, in double debug mode. These are the logs from when it finishes updating its Red Hat database until `wazuh-modulesd` daemon is Killed: 

```
2019/12/26 17:58:04 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2557 at wm_vuldet_check_feed(): INFO: (5494): The update of the Red Hat Enterprise Linux feed finished successfully.
2019/12/26 17:58:04 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2535 at wm_vuldet_check_feed(): DEBUG: (5461): Starting Wazuh CPE dictionary database update.
2019/12/26 17:58:04 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:3872 at wm_vuldet_fetch_wazuh_cpe(): DEBUG: (5457): Wazuh CPE dictionary OVAL is in its latest version. Update date: 2019-10-02T10:56Z
2019/12/26 17:58:04 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2559 at wm_vuldet_check_feed(): DEBUG: (5461): Starting Wazuh CPE dictionary database update.
2019/12/26 17:58:04 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2535 at wm_vuldet_check_feed(): DEBUG: (5461): Starting Microsoft Security Update database update.
2019/12/26 17:58:05 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2222 at wm_vuldet_index_feed(): DEBUG: (5458): Refreshing Microsoft Security Update databases.
2019/12/26 17:58:05 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:1245 at wm_vuldet_insert(): DEBUG: (5473): Inserting vulnerabilities.
2019/12/26 17:58:05 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:1272 at wm_vuldet_insert(): DEBUG: (5530): Inserting Microsoft Security Update dictonary.
2019/12/26 17:58:07 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2228 at wm_vuldet_index_feed(): DEBUG: (5459): Refresh of Microsoft Security Update database finished.
2019/12/26 17:58:07 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2233 at wm_vuldet_index_feed(): DEBUG: remove(tmp/vuln-temp): No such file or directory
2019/12/26 17:58:07 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2559 at wm_vuldet_check_feed(): DEBUG: (5461): Starting Microsoft Security Update database update.
2019/12/26 17:58:07 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:2533 at wm_vuldet_check_feed(): INFO: (5461): Starting National Vulnerability Database database update.
2019/12/26 17:58:07 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:1155 at wm_vuldet_update_feed(): DEBUG: (5512): Synchronizing the year 2010 of the vulnerability database.
2019/12/26 17:58:07 wazuh-modulesd:download[13440] wm_download.c:213 at wm_download_dispatch(): DEBUG: Downloading 'https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-2010.meta' to '/var/ossec/tmp/vuln-temp'
2019/12/26 17:58:07 wazuh-modulesd:download[13440] wm_download.c:228 at wm_download_dispatch(): DEBUG: Download of 'https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-2010.meta' finished.
2019/12/26 17:58:07 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector.c:1155 at wm_vuldet_update_feed(): DEBUG: (5512): Synchronizing the year 2011 of the vulnerability database.
2019/12/26 17:58:07 wazuh-modulesd:download[13440] wm_download.c:213 at wm_download_dispatch(): DEBUG: Downloading 'https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-2011.meta' to '/var/ossec/tmp/vuln-temp'
2019/12/26 17:58:08 wazuh-modulesd:download[13440] wm_download.c:228 at wm_download_dispatch(): DEBUG: Download of 'https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-2011.meta' finished.
2019/12/26 17:58:08 wazuh-modulesd:vulnerability-detector[13440] wm_vuln_detector_nvd.c:812 at wm_vuldet_fetch_nvd_cve(): DEBUG: (5460): National Vulnerability Database (2011) feed is outdated, so the update continues.
2019/12/26 17:58:08 wazuh-modulesd:download[13440] wm_download.c:213 at wm_download_dispatch(): DEBUG: Downloading 'https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-2011.json.gz' to '/var/ossec/tmp/req-3767575265'
2019/12/26 17:58:09 wazuh-modulesd:download[13440] wm_download.c:228 at wm_download_dispatch(): DEBUG: Download of 'https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-2011.json.gz' finished.
Killed
```

**Comment by vikman90** [comment]
## Configuration

This is the configuration that produced the problem:

```xml
  <vulnerability-detector>
    <enabled>yes</enabled>
    <interval>1m</interval>
    <ignore_time>6h</ignore_time>
    <run_on_start>yes</run_on_start>
    <provider name="canonical">
      <enabled>no</enabled>
      <os>precise</os>
      <os>trusty</os>
      <os>xenial</os>
      <os>bionic</os>
      <update_interval>1h</update_interval>
    </provider>
    <provider name="debian">
      <enabled>no</enabled>
      <os>wheezy</os>
      <os>stretch</os>
      <os>jessie</os>
      <os>buster</os>
      <update_interval>1h</update_interval>
    </provider>
    <provider name="redhat">
      <enabled>no</enabled>
      <update_from_year>2010</update_from_year>
      <update_interval>1h</update_interval>
    </provider>
    <provider name="nvd">
      <enabled>yes</enabled>
      <path>/root/nvdcve-1.1-2017.json.gz$</path>
      <update_interval>1m</update_interval>
    </provider>
  </vulnerability-detector>
```

**Comment by mikykeane** [comment]
I wanted to make a clarification, that even though that was the last configuration tested. The problem originated  with this configuration: 

```xml
<vulnerability-detector>
  <enabled>yes</enabled>
  <interval>5m</interval>
  <ignore_time>6h</ignore_time>
  <run_on_start>yes</run_on_start>
  <provider name="canonical">
    <enabled>no</enabled>
    <os>precise</os>
    <os>trusty</os>
    <os>xenial</os>
    <os>bionic</os>
    <update_interval>1h</update_interval>
  </provider>
  <provider name="debian">
    <enabled>no</enabled>
    <os>wheezy</os>
    <os>stretch</os>
    <os>jessie</os>
    <update_interval>1h</update_interval>
  </provider>
  <provider name="redhat">
    <enabled>no</enabled>
    <update_from_year>2010</update_from_year>
    <update_interval>1h</update_interval>
  </provider>
  <provider name="nvd">
    <enabled>yes</enabled>
    <update_from_year>2010</update_from_year>
    <update_interval>1h</update_interval>
  </provider>
</vulnerability-detector>
```

**Comment by jesusjimsa** [comment]
The problem appears to be caused by a high memory usage from the `cJSON` library when reading the `nvd` json file. 
When trying to parse it in the `json_fread` function using `cJSON_ParseWithOpts`, the system runs out of memory.  
Inside `cJSON_ParseWithOpts`, the program gets killed when executing `if (!parse_value(item, buffer_skip_whitespace(skip_utf8_bom(&buffer))))`.

**Comment by chemamartinez** [comment]
Merged #4427
